lib/Frontend/DiagnosticVerifier.cpp

```diff
@@ -437,40 +462,67 @@ DiagnosticVerifier::Result DiagnosticVerifier::verifyFile(unsigned BufferID) {
       size_t EndLoc = ExtraChecks.find("}}");
       if (EndLoc == StringRef::npos) {
         addError(ExtraChecks.data(),
-                 "didn't find '}}' to match '{{' in fix-it verification");
+                 "didn't find '}}' to match '{{' in diagnostic verification");
         break;
       }
       
       // Allow for close braces to appear in the replacement text.
       while (EndLoc+2 < ExtraChecks.size() && ExtraChecks[EndLoc+2] == '}')
         ++EndLoc;
-      
-      StringRef FixItStr = ExtraChecks.slice(2, EndLoc);
+
+      StringRef CheckStr = ExtraChecks.slice(2, EndLoc);
       // Check for matching a later "}}" on a different line.
-      if (FixItStr.find_first_of("\r\n") != StringRef::npos) {
+      if (CheckStr.find_first_of("\r\n") != StringRef::npos) {
         addError(ExtraChecks.data(), "didn't find '}}' to match '{{' in "
-                 "fix-it verification");
+                                     "diagnostic verification");
         break;
       }
-      
+
       // Prepare for the next round of checks.
       ExtraChecks = ExtraChecks.substr(EndLoc+2).ltrim();
-      
+
+      // If this check starts with 'educational-notes=', check for one or more
+      // educational notes instead of a fix-it.
+      if (CheckStr.startswith("educational-notes=")) {
+        if (Expected.EducationalNotes.hasValue()) {
+          addError(CheckStr.data(),
+                   "each verified diagnostic may only have one "
+                   "{{educational-notes=<#notes#>}} declaration");
+          continue;
+        }
+        StringRef NotesStr = CheckStr.substr(18); // Trim 'educational-notes='.
+        const auto startLoc =
+            NotesStr.data() - 20; // Include '{{educational-notes='
+        const auto endLoc =
```

Comment:
Doesn't `EndLoc` above have this value?

Reply:
`EndLoc` was an index at the end of the content in between the delimiters, so nearly the same thing. I've tried to rationalize some of the variable naming here and cut down on index manipulation. There's still follow up work that can be done, especially for the fix-it handling.